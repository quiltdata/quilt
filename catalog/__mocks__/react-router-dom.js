import { mockComponent } from 'testing/util';

export const Link = mockComponent('Link');
export const Redirect = mockComponent('Redirect');
