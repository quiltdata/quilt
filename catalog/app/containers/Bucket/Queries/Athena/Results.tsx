import * as React from 'react'
import * as M from '@material-ui/core'

import Perspective from 'components/Preview/renderers/Perspective'

import * as requests from '../requests'

function Empty() {
  return (
    <M.Paper>
      <M.Box p={3} textAlign="center">
        <M.Typography variant="h6">No results for this query</M.Typography>
        <M.Typography>
          Select another query execution or execute another query
        </M.Typography>
      </M.Box>
    </M.Paper>
  )
}

const useResultsStyles = M.makeStyles((t) => ({
  root: {
    padding: t.spacing(2),
  },
}))

interface ResultsProps {
  columns: requests.athena.QueryResultsColumns
  onLoadMore?: () => void
  rows: requests.athena.QueryResultsRows
}

export default function Results({ columns, rows, onLoadMore }: ResultsProps) {
  const classes = useResultsStyles()
  const data = React.useMemo(
    () =>
      rows.map((row) =>
        row.reduce(
          (memo, item, index) => ({
            ...memo,
            [columns[index]?.name || 'Unknown']: item,
          }),
          {},
        ),
      ),
    [columns, rows],
  )

  if (!data.length) return <Empty />

  return (
    <M.Paper className={classes.root}>
      <Perspective data={data} onLoadMore={onLoadMore} truncated={!!onLoadMore} />
    </M.Paper>
  )
}
