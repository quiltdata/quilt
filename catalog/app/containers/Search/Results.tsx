import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import { ES_REF_SYNTAX } from 'components/SearchResults'
import Skeleton from 'components/Skeleton'
import { useNavBar } from 'containers/NavBar'
import * as GQL from 'utils/GraphQL'
import StyledLink from 'utils/StyledLink'

import * as SearchUIModel from './model'

const useResultsSkeletonStyles = M.makeStyles((t) => ({
  root: {
    padding: t.spacing(2),
    '& + &': {
      marginTop: t.spacing(1),
    },
  },
  header: {
    display: 'flex',
  },
  expand: {
    marginLeft: 'auto',
    height: t.spacing(4),
    width: t.spacing(15),
  },
  title: {
    height: t.spacing(4),
    width: '55%',
  },
  version: {
    height: t.spacing(3.5),
    width: '75%',
    marginTop: t.spacing(1),
  },
  content: {
    height: t.spacing(30),
    marginTop: t.spacing(2),
  },
}))

interface ResultsSkeletonProps {
  className?: string
}

export function ResultsSkeleton({ className }: ResultsSkeletonProps) {
  const classes = useResultsSkeletonStyles()
  return (
    <div className={className}>
      <M.Paper className={classes.root}>
        <div className={classes.header}>
          <Skeleton className={classes.title} />
          <Skeleton className={classes.expand} />
        </div>
        <Skeleton className={classes.version} />
        <Skeleton className={classes.content} />
      </M.Paper>
      <M.Paper className={classes.root}>
        <div className={classes.header}>
          <Skeleton className={classes.title} />
          <Skeleton className={classes.expand} />
        </div>
        <Skeleton className={classes.version} />
        <Skeleton className={classes.content} />
      </M.Paper>
      <M.Paper className={classes.root}>
        <div className={classes.header}>
          <Skeleton className={classes.title} />
          <Skeleton className={classes.expand} />
        </div>
        <Skeleton className={classes.version} />
        <Skeleton className={classes.content} />
      </M.Paper>
    </div>
  )
}

const LABELS = {
  [SearchUIModel.ResultType.QuiltPackage]: 'packages',
  [SearchUIModel.ResultType.S3Object]: 'objects',
}

const useEmptyResultsStyles = M.makeStyles((t) => ({
  root: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
  },
  body: {
    maxWidth: '30rem',
  },
  list: {
    ...t.typography.body1,
    paddingLeft: 0,
  },
}))

interface EmptyResultsProps {
  className?: string
}

export function EmptyResults({ className }: EmptyResultsProps) {
  const classes = useEmptyResultsStyles()
  const {
    actions: { clearFilters, reset, setBuckets, setResultType },
    baseSearchQuery,
    state,
  } = SearchUIModel.use()
  const focus = useNavBar()?.focus

  const startNewSearch = React.useCallback(() => {
    reset()
    focus?.()
  }, [focus, reset])

  const resetBuckets = React.useCallback(() => {
    setBuckets([])
  }, [setBuckets])

  const otherResultType =
    state.resultType === SearchUIModel.ResultType.QuiltPackage
      ? SearchUIModel.ResultType.S3Object
      : SearchUIModel.ResultType.QuiltPackage

  const getTotalResults = (resultType: SearchUIModel.ResultType) =>
    GQL.fold(baseSearchQuery, {
      data: (data) => {
        const r =
          resultType === SearchUIModel.ResultType.QuiltPackage
            ? data.searchPackages
            : data.searchObjects
        switch (r.__typename) {
          case 'EmptySearchResultSet':
            return 0
          case 'ObjectsSearchResultSet':
          case 'PackagesSearchResultSet':
            return r.stats.total
          default:
            return null
        }
      },
      fetching: () => null,
      error: () => null,
    })

  const totalOtherResults = getTotalResults(otherResultType)

  const switchResultType = React.useCallback(() => {
    setResultType(otherResultType)
  }, [setResultType, otherResultType])

  let numFilters = state.filter.order.length
  if (state.resultType === SearchUIModel.ResultType.QuiltPackage) {
    numFilters += state.userMetaFilters.filters.size
  }

  return (
    <div className={cx(classes.root, className)}>
      <M.Typography variant="h4">No matching {LABELS[state.resultType]}</M.Typography>

      <M.Box mt={3} />
      <M.Typography variant="body1" align="center" className={classes.body}>
        Search for{' '}
        <StyledLink onClick={switchResultType}>{LABELS[otherResultType]}</StyledLink>{' '}
        instead{totalOtherResults != null && ` (${totalOtherResults} found)`} or adjust
        your search:
      </M.Typography>

      <ul className={classes.list}>
        {state.buckets.length > 0 && (
          <li>
            Search in <StyledLink onClick={resetBuckets}>all buckets</StyledLink>
          </li>
        )}
        {numFilters > 0 && (
          <li>
            Reset the <StyledLink onClick={clearFilters}>search filters</StyledLink>
          </li>
        )}
        <li>
          Edit your <StyledLink onClick={focus}>search query</StyledLink>
        </li>
        <li>
          Start <StyledLink onClick={startNewSearch}>from scratch</StyledLink>
        </li>
      </ul>
    </div>
  )
}

interface SearchErrorProps {
  className?: string
  kind?: 'unexpected' | 'syntax'
  details: React.ReactNode
}

export function SearchError({
  className,
  kind = 'unexpected',
  details,
}: SearchErrorProps) {
  const classes = useEmptyResultsStyles()
  const {
    actions: { reset },
  } = SearchUIModel.use()
  const focus = useNavBar()?.focus
  const startNewSearch = React.useCallback(
    (event) => {
      event.stopPropagation()
      reset()
      focus?.()
    },
    [reset, focus],
  )
  const tryAgain = React.useCallback(() => {
    // FIXME: retry GQL request
    window.location.reload()
  }, [])

  return (
    <div className={cx(classes.root, className)}>
      <M.Typography variant="h4">
        {kind === 'syntax' ? 'Query syntax error' : 'Unexpected error'}
      </M.Typography>
      <M.Box mt={3} />
      <M.Typography variant="body1" align="center" className={classes.body}>
        {kind === 'syntax' ? (
          <>
            Oops, couldn&apos;t parse that search.
            <br />
            Try quoting <StyledLink onClick={focus}>your query</StyledLink> or read about{' '}
            <StyledLink href={ES_REF_SYNTAX} target="_blank">
              supported query syntax
            </StyledLink>
            .
          </>
        ) : (
          <>
            Oops, something went wrong.
            <br />
            <StyledLink onClick={tryAgain}>Try again</StyledLink> or start a{' '}
            <StyledLink onClick={startNewSearch}>new search</StyledLink>.
          </>
        )}
      </M.Typography>

      <M.Box mt={3} />
      <M.Typography variant="h6">Error details</M.Typography>
      <M.Box mt={1} />
      <M.Typography variant="body2" className={classes.body}>
        {details}
      </M.Typography>
    </div>
  )
}
