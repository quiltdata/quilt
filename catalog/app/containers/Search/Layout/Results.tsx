import cx from 'classnames'
import * as React from 'react'
import * as RRDom from 'react-router-dom'
import * as M from '@material-ui/core'
import * as Icons from '@material-ui/icons'
import * as Lab from '@material-ui/lab'

import * as PD from 'containers/Bucket/PackageDialog'
import { useBucketStrict } from 'containers/Bucket/Routes'
import * as NamedRoutes from 'utils/NamedRoutes'
import assertNever from 'utils/assertNever'
import * as Format from 'utils/format'

import SortSelector from '../Sort'
import * as SearchUIModel from '../model'

import ColumnTitle from './ColumnTitle'

const useCreatePackageStyles = M.makeStyles({
  label: {
    display: 'block',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
  },
})

interface CreatePackageButtonProps {
  className: string
  onClick: () => void
}

function CreatePackageButton({ className, onClick }: CreatePackageButtonProps) {
  const classes = useCreatePackageStyles()
  const t = M.useTheme()
  const sm = M.useMediaQuery(t.breakpoints.down('sm'))
  const xs = M.useMediaQuery(t.breakpoints.down('xs'))
  return xs ? (
    <M.Button
      className={className}
      color="primary"
      onClick={onClick}
      size="medium"
      variant="contained"
    >
      <M.Icon>add</M.Icon>
    </M.Button>
  ) : (
    <M.Button
      className={className}
      color="primary"
      onClick={onClick}
      size={sm ? 'medium' : 'small'}
      startIcon={<M.Icon>add</M.Icon>}
      variant="contained"
    >
      <span className={classes.label}>Create new package</span>
    </M.Button>
  )
}

interface CreatePackageProps {
  className: string
}

function CreatePackage({ className }: CreatePackageProps) {
  const bucket = useBucketStrict()
  const dst = React.useMemo(() => ({ bucket }), [bucket])
  const { open, render } = PD.useCreateDialog({
    dst,
    delayHashing: true,
    disableStateDisplay: true,
  })
  return (
    <>
      <CreatePackageButton className={className} onClick={open} />
      {render({
        successTitle: 'Package created',
        successRenderMessage: ({ packageLink }) => (
          <>Package {packageLink} successfully created</>
        ),
        title: 'Create package',
      })}
    </>
  )
}

const I18_COUNT_RESULTS = {
  one: '1 result',
  other: (n: number) => (n > 0 ? `${n} results` : 'Results'),
}

const I18_COUNT_PACKAGES = {
  one: '1 package',
  // `-1` == secure search
  other: (n: number) => (n > 0 ? `${n} packages` : 'Packages'),
}

function resultsCountI18n(n: number, state: SearchUIModel.SearchUrlState) {
  if (
    state.resultType === SearchUIModel.ResultType.QuiltPackage &&
    state.view === SearchUIModel.View.Table
  ) {
    return Format.pluralify(n, I18_COUNT_PACKAGES)
  }
  return Format.pluralify(n, I18_COUNT_RESULTS)
}

function ResultsCount() {
  const model = SearchUIModel.use()
  const r = model.firstPageQuery
  switch (r._tag) {
    case 'fetching':
      return <Lab.Skeleton width={140} />
    case 'error':
      return null
    case 'data':
      switch (r.data.__typename) {
        case 'EmptySearchResultSet':
        case 'InvalidInput':
        case 'OperationError':
          return null
        case 'ObjectsSearchResultSet':
        case 'PackagesSearchResultSet':
          return <>{resultsCountI18n(r.data.total, model.state)}</>
        default:
          assertNever(r.data)
      }
    default:
      assertNever(r)
  }
}

const useFiltersButtonStyles = M.makeStyles((t) => ({
  root: {
    background: t.palette.background.paper,
  },
}))

interface FiltersButtonProps {
  className: string
  onClick: () => void
}

function FiltersButton({ className, onClick }: FiltersButtonProps) {
  const classes = useFiltersButtonStyles()
  return (
    <M.Button
      startIcon={<M.Icon fontSize="inherit">filter_list</M.Icon>}
      variant="outlined"
      className={cx(classes.root, className)}
      onClick={onClick}
    >
      Filters
    </M.Button>
  )
}

const useToggleButtonStyles = M.makeStyles({
  root: {
    // So, it occupies the same height as other buttons in that toolbar
    padding: '5px',
    // XXX:
    // This is a color hardcoded for MUI.Button#outlined
    // https://github.com/mui/material-ui/blob/v4.x/packages/material-ui/src/Button/Button.js#L70
    // Latest Lab.ToggleButtonGroup has correct color, but we don't use it
    borderColor: `rgba(0, 0, 0, 0.23)`,
  },
})

interface ToggleResultsViewProps {
  className: string
}

function ToggleResultsView({ className }: ToggleResultsViewProps) {
  const classes = useToggleButtonStyles()
  const model = SearchUIModel.use()
  const handleChange = React.useCallback(
    (_e, value: SearchUIModel.View) => model.actions.setView(value),
    [model.actions],
  )
  return (
    <Lab.ToggleButtonGroup
      value={model.state.view}
      className={className}
      exclusive
      onChange={handleChange}
      size="small"
    >
      <Lab.ToggleButton value={SearchUIModel.View.Table} classes={classes}>
        <Icons.GridOn />
      </Lab.ToggleButton>
      <Lab.ToggleButton value={SearchUIModel.View.List} classes={classes}>
        <Icons.List />
      </Lab.ToggleButton>
    </Lab.ToggleButtonGroup>
  )
}

const useResultsStyles = M.makeStyles((t) => ({
  root: {
    display: 'flex',
    flexDirection: 'row',
    minHeight: t.spacing(4.5),
    flexWrap: 'wrap',
    [t.breakpoints.down('xs')]: {
      flexDirection: 'column',
    },
  },
  button: {
    '& + &': {
      marginLeft: t.spacing(1),
    },
  },
  controls: {
    alignItems: 'center',
    display: 'flex',
    flexGrow: 1,
    overflow: 'hidden',
  },
  title: {
    flexShrink: 0,
    marginRight: t.spacing(2),
  },
  create: {
    marginRight: t.spacing(4),
  },
  controlsInner: {
    alignItems: 'center',
    display: 'flex',
    flexGrow: 1,
    flexShrink: 0,
    justifyContent: 'flex-end',
    whiteSpace: 'nowrap',
  },
}))

interface ResultsProps {
  onFilters?: () => void
}

export default function Results({ onFilters }: ResultsProps) {
  const model = SearchUIModel.use()
  const r = model.firstPageQuery
  const classes = useResultsStyles()
  const { paths } = NamedRoutes.use()
  return (
    <div className={classes.root}>
      <ColumnTitle className={classes.title}>
        <ResultsCount />
      </ColumnTitle>

      <div className={classes.controls}>
        <RRDom.Switch>
          <RRDom.Route path={paths.bucketRoot}>
            {r._tag === 'data' &&
              (r.data.__typename === 'ObjectsSearchResultSet' ||
                r.data.__typename === 'PackagesSearchResultSet') && (
                <CreatePackage className={classes.create} />
              )}
          </RRDom.Route>
        </RRDom.Switch>

        <div className={classes.controlsInner}>
          {model.state.resultType === SearchUIModel.ResultType.QuiltPackage && (
            <ToggleResultsView className={classes.button} />
          )}
          {onFilters && <FiltersButton className={classes.button} onClick={onFilters} />}
          <SortSelector className={classes.button} />
        </div>
      </div>
    </div>
  )
}
