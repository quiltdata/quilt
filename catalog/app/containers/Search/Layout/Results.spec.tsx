import * as React from 'react'
import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'

import Results from './Results'

jest.mock('@material-ui/core', () => ({
  Button: ({ children }: React.PropsWithChildren<{}>) => <button>{children}</button>,
  Icon: ({ children }: React.PropsWithChildren<{}>) => <span>{children}</span>,
  makeStyles: () => () => ({}),
  useTheme: () => ({
    breakpoints: { down: () => false },
    spacing: (x: number) => x * 8,
  }),
  useMediaQuery: () => false,
}))

jest.mock('@material-ui/icons', () => ({
  GridOn: () => 'table icon',
  List: () => 'list icon',
}))

jest.mock('@material-ui/lab', () => ({
  Skeleton: () => <div>Loadingâ€¦</div>,
  ToggleButtonGroup: ({
    value,
    children,
  }: React.PropsWithChildren<{ value: string }>) => (
    <ul data-selected={value}>{children}</ul>
  ),
  ToggleButton: ({ children, value }: React.PropsWithChildren<{ value: string }>) => (
    <li data-value={value}>{children}</li>
  ),
}))

const model = {
  state: {
    resultType: 'p', // QuiltPackage
    view: 'l', // List
    searchString: 'test',
    buckets: ['test-bucket'],
    order: 'BEST_MATCH',
    filter: {
      predicates: {},
      order: [],
    },
    userMetaFilters: {
      filters: new Map(),
    },
    latestOnly: true,
  },
  actions: {
    setView: jest.fn(),
    setOrder: jest.fn(),
  },
  firstPageQuery: {
    _tag: 'fetching',
  } as any,
  baseSearchQuery: {
    _tag: 'fetching',
  } as any,
}

jest.mock('../model', () => ({
  use: () => model,
  ResultType: {
    QuiltPackage: 'p',
    S3Object: 'o',
  },
  View: {
    Table: 't',
    List: 'l',
  },
}))

jest.mock('containers/Bucket/PackageDialog/PackageCreationForm', () => ({
  usePackageCreationDialog: () => ({
    open: jest.fn(),
    render: () => <>Don't forget to render dialog</>,
  }),
}))

jest.mock('containers/Bucket/Routes', () => ({
  useBucketStrict: () => 'test-bucket',
}))

jest.mock('./ColumnTitle', () => ({ children }: React.PropsWithChildren<{}>) => (
  <div>Column Title: {children}</div>
))

jest.mock('../Sort', () => () => <div>Sort Selector</div>)

jest.mock('utils/NamedRoutes', () => ({
  use: () => ({
    paths: {
      bucketRoot: '/b/:bucket',
    },
  }),
}))

describe('containers/Search/Layout/Results', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    model.firstPageQuery = { _tag: 'fetching' }
    model.state.resultType = 'p'
  })

  it('renders with loading state', () => {
    const { container } = render(
      <MemoryRouter>
        <Results />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })

  it('renders with data and shows number of results', () => {
    model.firstPageQuery = {
      _tag: 'data',
      data: {
        __typename: 'PackagesSearchResultSet',
        total: 5,
      },
    }

    const { container } = render(
      <MemoryRouter>
        <Results />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })

  it('renders with FiltersButton when onFilters prop is provided', () => {
    model.firstPageQuery = {
      _tag: 'data',
      data: {
        __typename: 'PackagesSearchResultSet',
        total: 3,
      },
    }

    const { container } = render(
      <MemoryRouter>
        <Results onFilters={jest.fn()} />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })

  it('does not show ToggleResultsView for S3Object result type', () => {
    model.state.resultType = 'o' // S3Object
    model.firstPageQuery = {
      _tag: 'data',
      data: {
        __typename: 'ObjectsSearchResultSet',
        total: 5,
      },
    }

    const { container } = render(
      <MemoryRouter>
        <Results />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })

  it('renders error state', () => {
    model.firstPageQuery = {
      _tag: 'error',
      error: new Error('Test error'),
    }

    const { container } = render(
      <MemoryRouter>
        <Results />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })

  it('shows Create Package button in bucket', () => {
    model.firstPageQuery = {
      _tag: 'data',
      data: {
        __typename: 'PackagesSearchResultSet',
        total: 5,
      },
    }
    const { container } = render(
      <MemoryRouter initialEntries={['/b/test-bucket/packages/my-package']}>
        <Results />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })
})
