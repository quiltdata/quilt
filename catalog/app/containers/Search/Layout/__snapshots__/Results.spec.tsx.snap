// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Search/Layout/Results does not show ToggleResultsView for S3Object result type 1`] = `
<div>
  <div>
    <div>
      Column Title: 
      5 results
    </div>
    <div>
      <div>
        <div>
          Sort Selector
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/Search/Layout/Results renders error state 1`] = `
<div>
  <div>
    <div>
      Column Title: 
    </div>
    <div>
      <div>
        <ul
          data-selected="l"
        >
          <li
            data-value="t"
          >
            table icon
          </li>
          <li
            data-value="l"
          >
            list icon
          </li>
        </ul>
        <div>
          Sort Selector
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/Search/Layout/Results renders with FiltersButton when onFilters prop is provided 1`] = `
<div>
  <div>
    <div>
      Column Title: 
      3 results
    </div>
    <div>
      <div>
        <ul
          data-selected="l"
        >
          <li
            data-value="t"
          >
            table icon
          </li>
          <li
            data-value="l"
          >
            list icon
          </li>
        </ul>
        <button>
          Filters
        </button>
        <div>
          Sort Selector
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/Search/Layout/Results renders with data and shows number of results 1`] = `
<div>
  <div>
    <div>
      Column Title: 
      5 results
    </div>
    <div>
      <div>
        <ul
          data-selected="l"
        >
          <li
            data-value="t"
          >
            table icon
          </li>
          <li
            data-value="l"
          >
            list icon
          </li>
        </ul>
        <div>
          Sort Selector
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/Search/Layout/Results renders with loading state 1`] = `
<div>
  <div>
    <div>
      Column Title: 
      <div>
        Loadingâ€¦
      </div>
    </div>
    <div>
      <div>
        <ul
          data-selected="l"
        >
          <li
            data-value="t"
          >
            table icon
          </li>
          <li
            data-value="l"
          >
            list icon
          </li>
        </ul>
        <div>
          Sort Selector
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/Search/Layout/Results shows Create Package button in bucket 1`] = `
<div>
  <div>
    <div>
      Column Title: 
      5 results
    </div>
    <div>
      <button>
        <span>
          Create new package
        </span>
      </button>
      Don't forget to render dialog
      <div>
        <ul
          data-selected="l"
        >
          <li
            data-value="t"
          >
            table icon
          </li>
          <li
            data-value="l"
          >
            list icon
          </li>
        </ul>
        <div>
          Sort Selector
        </div>
      </div>
    </div>
  </div>
</div>
`;
